rosbag2_py/src/rosbag2_py/_storage.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_storage/storage_filter.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+PYBIND11_MODULE(_storage, m) {
+  m.doc() = "Python wrapper of the rosbag2 utilities API";
+
+  pybind11::class_<rosbag2_cpp::ConverterOptions>(m, "ConverterOptions")
+  .def(pybind11::init())
+  .def_readwrite(
+    "input_serialization_format",
+    &rosbag2_cpp::ConverterOptions::input_serialization_format)
+  .def_readwrite(
+    "output_serialization_format",
+    &rosbag2_cpp::ConverterOptions::output_serialization_format);
+
+  pybind11::class_<rosbag2_cpp::StorageOptions>(m, "StorageOptions")
+  .def(pybind11::init())
```

Comment:
Are we able to bind this constructor to a "brace initializer" (https://pybind11.readthedocs.io/en/stable/advanced/classes.html#brace-initialization)? If not, can we define factory functions here in the python wrapper? I'd like for the py-side objects to be single-call constructable: e.g. instead of 

```
   storage_options = rosbag2_py.StorageOptions()
   storage_options.uri = path
   storage_options.storage_id = 'sqlite3'
```

the API would be better if we can

```
storage_options = rosbag2_py.StorageOptions(uri=path, storage_id='sqlite3')
```

This comment/question applies to all the POD structs exposed in this PR (ConverterOptions, StorageOptions, StorageFilter, TopicMetadata i think is all of them)

Reply:
I tried out the brace initialization and found that it doesn't let us have keyword arguments. Some of the structs have default values like 0, and currently with the pybind brace initialization alone, I have to initialize it like this, which is kind of dumb:
```
storage_options = rosbag2_py.StorageOptions(path, 'sqlite3', 0, 0, 0)
```
The three 0s are max_bagfile_size, max_bagfile_duration, max_cache_size, which are by default 0s in C++. With the argument types specified in `<>` in the brace initialization, they aren't optional.

I think the keyword arguments might do what we want: https://pybind11.readthedocs.io/en/stable/advanced/functions.html#accepting-args-and-kwargs
I can look into this next week.

Reply:
Maybe we can do "named arguments" with defaults like the following? This may need us to define constructors or factory functions, which is maybe not ideal.

```
someclass
    .def(pybind11::init<long, double, double>(),
            pybind11::arg("arg1")=0, pybind11::arg("arg2")=0., pybind11::arg("arg3")=1.
    )
```

If it gets too difficult, maybe we do end up keeping it as a "default then set" struct - I don't love it as an API but it may be more maintainable

Reply:
btw: It is sufficient to pass converter options without anything set:

```python
        storage_options = rosbag2.StorageOptions()
        storage_options.uri = uri
        storage_options.storage_id = 'sqlite3'
        reader.open(storage_options, rosbag2.ConverterOptions())
```

Reply:
@emersonknapp Nice find! That's a lot simpler than the `kwargs` and works. 0603d27